app/Http/Controllers/Rest/UsersController.php

```diff
@@ -0,0 +1,152 @@
+<?php
+/**
+ * Copyright (c) 2018 Adshares sp. z o.o.
+ *
+ * This file is part of AdServer
+ *
+ * AdServer is free software: you can redistribute and/or modify it
+ * under the terms of the GNU General Public License as published
+ * by the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * AdServer is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty
+ * of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with AdServer. If not, see <https://www.gnu.org/licenses/>
+ */
+
+namespace Adshares\Adserver\Http\Controllers\Rest;
+
+use Adshares\Adserver\Http\Controllers\AppController;
+use Adshares\Adserver\Mail\UserEmailActivate;
+use Adshares\Adserver\Models\Token;
+use Adshares\Adserver\Models\User;
+use Illuminate\Http\Request;
+use Illuminate\Support\Facades\Auth;
+use Illuminate\Support\Facades\DB;
+use Illuminate\Support\Facades\Mail;
+use Illuminate\Support\Facades\Validator;
+use JsonRPC\Exception\ServerErrorException;
+
+class UsersController extends AppController
+{
+    protected $email_activation_token_time = 24 * 60 * 60; // 24 hours
+    protected $email_activation_resend_limit = 15 * 60; // 15 minutes
+    protected $email_change_token_time = 60 * 60; // 1 hour
+    protected $email_new_change_resend_limit = 5 * 60; // 1 minute
+
+    /**
+     * Create a new controller instance.
+     */
+    public function __construct()
+    {
+        $this->middleware('snake_casing')->except(['emailChangeStep1']);
+    }
+
+    public function add(Request $request)
+    {
+        $this->validateRequestObject($request, 'user', User::$rules_add);
+        Validator::make($request->all(), ['uri' => 'required'])->validate();
+
+        DB::beginTransaction();
+        $user = User::register($request->input('user'));
+        Mail::to($user)->queue(
+            new UserEmailActivate(
+                Token::generate('email-activate', $this->email_activation_token_time, $user->id),
+                $request->input('uri')
+            )
+        );
+        DB::commit();
+
+        return self::json($user->toArray(), 201)
+            ->header('Location', route('app.users.read', ['user_id' => $user->id]));
+    }
+
+    public function count(Request $request)
+    {
+        throw new ServerErrorException('Not implemented yet');
+    }
+
+    public function browse(Request $request)
+    {
+        return self::json([], 501, ['message' => 'not yet implemented <3']);
```

Comment:
![CRITICAL](https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL') Define a constant instead of duplicating this literal "message" 4 times. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=php%3AS1192)
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Remove the code after this "return". [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=php%3AS1763)